### ISSUE TITLE
Failed to run grpc python on multiprocessing

### ISSUE THREAD

**Issue by TennyZhuang** [grpc/grpc#18321]
### What version of gRPC and what language are you using?
 
```
pip3 show grpcio
Name: grpcio
Version: 1.19.0
Summary: HTTP/2-based RPC framework
Home-page: https://grpc.io
Author: The gRPC Authors
Author-email: grpc-io@googlegroups.com
License: Apache License 2.0
Location: /usr/local/lib/python3.7/site-packages
Requires: six
```
 
### What operating system (Linux, Windows,...) and version?
 
macOS 10.14
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 
Python 3.7.2
 
### What did you do?
If possible, provide a recipe for reproducing the error. Try being specific and include code snippets if helpful.
 
I run route_guide example and modify it on multiprocessing.

All i have modified is `route_guide_client.py`

The modified route_guide_client can be found here https://paste.ubuntu.com/p/nxvJbzXXyq/

I failed to using one channel on multi processes before `fork()` firstly, then I tried per process per connection, and it still failed.

```python3
from multiprocessing import Process, Manager

m = Manager()
s = m.dict()
stub = None

def get_stub():
    global stub
    import os
    pid = os.getpid()
    if pid in s:
        return stub

    print('init stub for ', pid)
    channel = grpc.insecure_channel('localhost:50051')
    stub = route_guide_pb2_grpc.RouteGuideStub(channel)
    s[pid] = 1
    return stub
```

### What did you expect to see?

Like everything in example
 
### What did you see instead?
 
Blocking  on first  gRPC  request  after `fork()`
 
### Anything else we should know about your project / environment?

I want to write a library use gRPC Python, and I cannot control when `fork` is called by my user, so multi process support is required.

I found multiple issues about this problem, but all closed, and i try on the latest  grpcio  version but still failed, so I create a new issue.

**Comment by gnossen** [comment]
It maybe isn't documented quite well enough in [this doc](https://github.com/grpc/grpc/blob/master/doc/fork_support.md#111), but [the PR that implemented client-side fork support](https://github.com/grpc/grpc/pull/16264) is quite clear on the matter:

> ### gRPC Core's API for fork support
> 
> A process may fork after invoking grpc_init() and use gRPC in the child if and only if the child process first destroys all gRPC resources inherited from the parent process and invokes grpc_shutdown().
> Subsequent to this, the child will be able to re-initialize and use gRPC. After fork, the parent process will be able to continue to use existing gRPC resources such as channels and calls without interference
> from the child process.
> 
> ### gRPC Python behavior at fork()
> 
> To facilitate gRPC Python applications meeting the above constraints, gRPC Python will automatically destroy and shutdown all gRPC Core resources in the child's post-fork handler, including cancelling in-flight calls. From the client's perspective, the child process is now free to create new channels and use gRPC.


The code you've linked never closes the channel it opens in the parent process, so the child processes are susceptible to deadlocks and hangs. I was able to make your linked example code work simply by closing the channel created in the parent process by the call to `f` in the `run` function. I'd recommend you use the [context manager idiom](https://github.com/grpc/grpc/blob/master/examples/python/helloworld/greeter_client.py#L29) every time you instantiate a channel.

Closing this issue for now. @TennyZhuang If this answer does not satisfy you, then please feel free to reopen.

**Comment by TennyZhuang** [comment]
Thanks for your response, and it works if I always close the channel.
But I create a library using gRPC and pack it to Python SDK use gRPC-python. And it is overhead if I always create a new channel on every connection.
I do a workaround that patch the `os.fork` function and always close the channel before the really `fork` called and recreate the connection after fork. This is my code:

```python
def getController():
    channel = grpc.insecure_channel('HOST')
    stub = XXXControllerStub(channel)

    def before():
        nonlocal channel
        channel.close()

    def after():
        nonlocal channel, stub
        channel = grpc.insecure_channel('HOST')
        stub = XXXControllerStub(channel)

    if hasattr(os, 'register_at_fork'):
        # https://docs.python.org/3.7/library/os.html#os.register_at_fork
        # features for Python 3.7
        os.register_at_fork(before=before, after_in_parent=after, after_in_child=after)
    else:
        orig_fork = os.fork

        @functools.wraps(os.fork)
        def new_fork():
            before()
            pid = orig_fork()
            after()
            return pid

        os.fork = new_fork

    global getController

    def getController():
        return stub
    return stub
```

A better solution is use [pthread_atfork](http://pubs.opengroup.org/onlinepubs/007904975/functions/pthread_atfork.html).
I found that in the PR which your mentioned, there is already some code that call `pthread_atfork`, and it seems that `channel._close_on_fork` will be called automatically in [postfork](https://github.com/grpc/grpc/pull/16264/files#diff-a0f10ee5af06201f5f2ca08f05f4a802R72).
Is it possible that do all the things in gRPC library? In pseudocode:

```
pre_fork:
Close event loop, close all active channels and record them.

post_fork:
Reconnect all channels recorded.
```

Then all things about `fork` will be shielded for gRPC users.

**Comment by TennyZhuang** [comment]
It seems that there is no reopen button
@gnossen

**Comment by TennyZhuang** [comment]
And here is another error if I use python 3.5 and use that `get_controller` before and after fork().

```
Exception ignored in: <bound method Channel.__del__ of <grpc._channel.Channel object at 0x7f1720d07ef0>>
Traceback (most recent call last):
  File "/home/USER/.local/lib/python3.5/site-packages/grpc/_channel.py", line 1063, in __del__
TypeError: 'NoneType' object is not callable
```

grpcio Version: 1.18.0

**Comment by gnossen** [comment]
@TennyZhuang 
Going back to your first example, can set the environment variable `GRPC_ENABLE_FORK_SUPPORT=1`. I believe the program should terminate without the need for a custom `at_fork` handler.

**Comment by somerska** [comment]
@gnossen 
nowhere is it documented how to set an environment variable for grpc python on windows...I've tried many things without success.
